aten/src/ATen/native/cuda/CUDALoops.cuh

```diff
@@ -144,95 +146,98 @@ namespace modern {
 
 namespace detail {
 
-template <typename func_t, typename array_t, std::size_t... I>
-__device__ inline constexpr decltype(auto) invoke_with_array_impl(func_t f, array_t t, std::index_sequence<I...>)
-{
-    return f(t[I]...);
-}
-template <typename func_t, typename array_t>
-__device__ inline constexpr decltype(auto) invoke_with_array(func_t f, array_t a) {
-  constexpr auto arity = function_traits<func_t>::arity;
-  return invoke_with_array_impl(f, a, std::make_index_sequence<arity>{});
-}
+// The `pointers` converts std::tuple<T1, T2, ....> to std::tuple<T1*, T2*, ....>
 
-namespace arg_type {
+template <typename T>
+struct pointers_helper {};
 
-// We need a way to compute the argument type of a function. But
-// for nullary function, it does not really have an argument type
-// in this case, we still need to return a valid type, but we don't
-// really care what type this is.
+template <typename... types>
+struct pointers_helper<std::tuple<types...>> {
+  using type = std::tuple<types *...>;
+};
 
-struct dont_care {};
+template <typename T>
+using pointers = typename pointers_helper<T>::type;
 
-template <typename func_t, std::size_t arity>
-struct arg_type_helper {
-  using type = typename function_traits<func_t>::template arg<0>::type;
+template<template<int i> typename func, int end, int current=0>
```

Comment:
comment here about why you need static unroll and what it does will be useful (you have brief description in the PR, but it should also be in the code)